Replace boolbv_mapt::map_bitt by bvt

tautschnig opened the pull request:
The Boolean flag is_set does not need to be stored for each individual
bit as all entries will either be false (initial state) or be true. It
is not possible for only some of the bits to be set.

On the benchmark that triggered me looking into this, memory consumption
caused by boolbv_mapt::get_map_entry was 4.50 GB, and is now 2.25 GB
(which is reasonable, because the bool-typed struct member likely
resulted in padding being generated that would effectively double the
memory consumption).

In addition to memory savings, some of the code can now be simplified as
the entire vector can be copied at once.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Replace boolbv_mapt::map_bitt by bvt

The Boolean flag is_set does not need to be stored for each individual
bit as all entries will either be false (initial state) or be true. It
is not possible for only some of the bits to be set.

On the benchmark that triggered me looking into this, memory consumption
caused by boolbv_mapt::get_map_entry was 4.50 GB, and is now 2.25 GB
(which is reasonable, because the bool-typed struct member likely
resulted in padding being generated that would effectively double the
memory consumption).

In addition to memory savings, some of the code can now be simplified as
the entire vector can be copied at once.

martin-cs reviewed (approved):
Seems reasonable.

tautschnig commented:
Thanks @martin-cs @peterschrammel for your approvals! @kroening I believe you will have authored this code originally and might have extra insight as to why having an `is_set` for each bit was the right thing to do. I'll thus hold off on merging this for a moment to give @kroening a chance to comment as well.

hannes-steffenhagen-diffblue commented on src/solvers/flattening/boolbv_map.cpp:
```diff
@@ -20,27 +20,18 @@ Author: Daniel Kroening, kroening@kroening.com
-    }
-
-    result=result+ch;
+    result += (value.is_true() ? '1' : (value.is_false() ? '0' : '?'));
```
With the way this is written right now I'd expect something like `10?001` to be possible result, but from the commit message it should either be all `?`s or none of them should be right?

tautschnig replied:
Not quite: the bits not being set would be printed as a string of '*'. The '?' refers to an individual variable not having a value under a particular assignment computed by a solver. That said: it seems the above is dead code, so perhaps the distinction between `is_set` (which perhaps should be "is_unconstrained" (!)) and is-not-set-under-an-assignment isn't all that important.

hannes-steffenhagen-diffblue commented on src/solvers/flattening/bv_dimacs.cpp:
```diff
@@ -49,21 +49,24 @@ bool bv_dimacst::write_dimacs(std::ostream &out)
+    {
+      out << ' ';
+
+      if(!m.second.is_set)
```
Same thing here, I guess. With the changes to me it looks like now the case where the whole map is not set is "separate" logically.

tautschnig replied:
I first wanted to keep the changes to a minimum, but ultimately that wasn't possible anyway. I have now clarified the code structure.

hannes-steffenhagen-diffblue reviewed (commented):
Don't know enough about this to be confident here, just leave my 2c.

tautschnig committed:
Add a regression test of DIMACS output

We've had this functionality for a long time, but seemingly never added
a test for it.

tautschnig committed:
boolbv_mapt: only create entries via {get,set}_literals

This is a first step towards getting rid of is_set altogether, making
sure that get_literals/set_literals are the only way to create entries
in the map. This required adding accessor functions to ensure the
mapping member cannot be externally modified. Calls to the previous
get_map_entry method were replaced by either using get_literals or
inlining the code.

tautschnig committed:
Remove boolbv_mapt::map_entryt::is_set

This is always true for any existing entry.

tautschnig committed:
Remove boolbv_mapt::map_entryt::{bvtype,width}

The bvtype member was unused and the width can always be computed from
existing information.

tautschnig committed:
Make boolbv_mapt::get_literals return a bvt

This avoids callers creating a bvt just so that get_literals can copy
its own bvt into the caller's bvt.

tautschnig committed:
Remove 'unknown' parameter from bv_get_rec

The entries in the vector are all guaranteed to be false. Thus, any
branches depending on the value can be cleaned up, and the vector no
longer needs to be passed around.

tautschnig commented:
@kroening I have extended this as discussed, which really permitted a number of further simplifications.
